Add a kpet.run class for test cases

emanchado opened the pull request:
Other "data" classes have a simplified version in kpet.run. This is
the version that is passed to the templates, and only contain
attributes that are safe to put in templates. The exception for that
was Case, so create a kpet.run.Case class to make sure we only expose
the things we want.

Ref. FASTMOVING-1443

emanchado committed:
Add a kpet.run class for test cases

Other "data" classes have a simplified version in kpet.run. This is
the version that is passed to the templates, and only contain
attributes that are safe to put in templates. The exception for that
was Case, so create a kpet.run.Case class to make sure we only expose
the things we want. Also pass it to the kpet.run.Suite constructor,
which now excepts a list of those objects.

Ref. FASTMOVING-1443

emanchado committed:
Remove references to url_suffix from documentation

The url_suffix attribute was removed some time ago, but the
documentation was never updated. So, do that.

emanchado commented on kpet/run.py:
```diff
@@ -18,6 +18,18 @@
+
+    def __init__(self, case):
+        for attr in self.EXPORTED_ATTRIBUTES:
+            setattr(self, attr, getattr(case, attr))
```
I don't like having lots of assignments when there isn't any transformation, so I did this loop instead. It makes it completely different to the other `kpet.run` classes, so I don't know if it's worth it, but hey.

spbnick replied:
That's OK, but can you maybe put the list of attributes into a local variable inside `__init__()`? We don't really need to expose it, or keep it inside the class. Better prevent interface creep.

Also, if you're going to do this, could you precede this commit with another one doing the same for `kpet.run.Suite`?

emanchado commented on kpet/run.py:
```diff
@@ -32,8 +44,9 @@ def __init__(self, suite, cases):
+        suite_case_names = [c.name for c in suite.cases]
         for case in cases:
-            assert case in suite.cases
+            assert case.name in suite_case_names
```
As `cases` is not a list of `kpet.data.Case` anymore, but `kpet.run.Case`, we cannot assert like before. I don't know if the `name` is good enough or if this could fail in some circumstances.

Another solution could be making this receive a list of `kpet.data.Case`, and make the conversion inside this method (in the `self.cases = cases` below)... but I'm not sure I like it so much either, plus the `Host` constructor does receive a list of `kpet.run.Suite` objects, not a list of `kpet.data.Suite` objects. So I'm consistent with the pattern used in other places.

What do you think?

spbnick replied:
I think let's simply drop the assertion for now. It's becoming too complicated.

I'm now thinking if we could do away with `kpet.run.Suite` altogether and just have a flat list of test case run objects right under `kpet.run.Host`, with properties from both `Suite` and `Case` aggregated (and some aggregated into `Host`). This might make it easier to write templates, randomize execution order, and other things. So let's not try too hard here.

emanchado reviewed (commented):
Some extra comments to give some insight as to why I implemented things this way. Not sure it's great, but I'm not super fond of the alternative, either, so I'm not sure what's best.

spbnick commented on README.md:
```diff
@@ -91,8 +91,6 @@ Each suite is an object with the following attributes:
 * `maintainers`: List of strings with the names and emails of the test
   maintainers.
-* `url_suffix`: URL suffix to use in the Beaker task fetch URL,
-  ie. the bit after the "#".
```
Ah, good catch, thanks a lot!

spbnick commented on kpet/run.py:
```diff
@@ -18,6 +18,18 @@
+                           "partitions", "kickstart", "waived", "role",
+                           "environment", "maintainers"]
+
+    def __init__(self, case):
```
Please document the constructor.

spbnick commented on kpet/run.py:
```diff
@@ -18,6 +18,18 @@
 
+class Case:
+    # pylint: disable=too-few-public-methods,too-many-instance-attributes
+    """A test case in a test suite"""
```
This should say "A test case run". As distinct from a "test case", that is `kpet.data.Case`. And also let's not restrict the case to only belonging to a test suite.

spbnick commented on kpet/run.py:
```diff
@@ -32,8 +44,9 @@ def __init__(self, suite, cases):
         """
```
Oh, and please update the documentation to reflect the type of `cases` argument.

emanchado committed:
Simplify the kpet.run.Suite constructor

Instead of doing lots of attribute copying one by one, make a list and
iterate through it.

spbnick reviewed (approved):
Looks good to me! Thank you, Esteban :)

@cki_bot, please test!
